### ISSUE TITLE
Custom types - getting "ERROR: type 'x' already exists" in multiple cases

### ISSUE THREAD

**Issue by nguiard** [pgcentralfoundation/pgrx#1076]
Hi,
Thanks for this awesome library.

I'm fooling around with custom types.

- This works:
```rust
use pgx::prelude::*;
use serde::{Serialize, Deserialize};

pgx::pg_module_magic!();


#[derive(Debug, Serialize, Deserialize, PostgresType)]
pub struct A;

#[pg_extern]
fn myfunc() -> A {
    A
}
```

- However, this does not anymore, when you wrap the return in a `Result`
```rust
use pgx::prelude::*;
use pgx::spi;
use serde::{Serialize, Deserialize};

pgx::pg_module_magic!();


#[derive(Debug, Serialize, Deserialize, PostgresType)]
pub struct A;

#[pg_extern]
fn myfunc() -> Result<A, spi::Error> {
    Ok(A)
}
```
It compiles fine of course, but when you try to `create extension`, you get `ERROR:  type "a" already exists`. (also, it's not dependent on `spi`, changing `spi::Error` to `std::error::Error` produces the same result).

It seems that this is tied to ordering problems in the generated sql file. Because this also fails with the same error:
```rust
use pgx::prelude::*;
use pgx::spi;
use serde::{Serialize, Deserialize};

pgx::pg_module_magic!();


#[derive(Debug, Serialize, Deserialize, PostgresType)]
pub struct B;

#[derive(Debug, Serialize, Deserialize, PostgresType)]
pub struct A;

#[pg_extern]
fn myfunc(_x: B) -> Result<A, spi::Error> {
    Ok(A)
}
```

But, rename `struct A` to `struct Z`, and it now works!
```rust
use pgx::prelude::*;
use pgx::spi;
use serde::{Serialize, Deserialize};

pgx::pg_module_magic!();


#[derive(Debug, Serialize, Deserialize, PostgresType)]
pub struct B;

#[derive(Debug, Serialize, Deserialize, PostgresType)]
pub struct Z;

#[pg_extern]
fn myfunc(_x: B) -> Result<Z, spi::Error> {
    Ok(Z)
}
```

**Comment by marcusllorente-msft** [comment]
Hi, I am encountering the same error as well. @nguiard I don't see why renaming the struct from A to Z would work, and I can't seem to make it work on my end when replicating the change from `struct A` to `struct Z`. Any insight?

**Comment by eeeebbbbrrrr** [comment]
@marcusllorente-msft which pgrx version?  Have you tried with the v0.10.0 betas?

**Comment by marcusllorente-msft** [comment]
@eeeebbbbrrrr I'm using crate v0.9.8. I tried following the steps to get the beta, but I am getting this error: ``pgrx-0.9.8` shouldn't be used with `cargo-pgrx-0.10.0-beta.0`, please use `pgrx = "~0.10.0-beta.0"` in your Cargo.toml`.`` and my Cargo.toml file includes this line: `pgrx = "~0.10.0-beta.0"` in the dependencies section. I am new to PGRX so apologies if this is a silly question

**Comment by eeeebbbbrrrr** [comment]
oh, yeah, you'd need to install the beta version of cargo-pgrx directly too:  `cargo install cargo-pgrx --version 0.10.0-beta.1` --locked`

Note that I typed `beta.1` here, so you'd update your Cargo.toml as well.

I don't know for certain if this is fixed in the 0.10.0 betas but there was some work around better understanding `Result` return types.  We haven't had a chance to actually look at this issue specifically.

If beta.1 doesn't help here then I'll commit to look at this directly next week.

**Comment by marcusllorente-msft** [comment]
@eeeebbbbrrrr Just tried installing both `beta.1` and `beta.0`, both seem to raise errors when I include the correct pgrx version:
`   0: `pgrx-0.9.8` shouldn't be used with `cargo-pgrx-0.10.0-beta.1`, please use `pgrx = "~0.10.0-beta.1"` in your `Cargo.toml`.`, perhaps I am missing something? I did the `cargo install` and `cargo pgrx init` commands too.

**Comment by eeeebbbbrrrr** [comment]
It seems like you've got `cargo-pgrx` installed correctly and now need to update your extension's `Cargo.toml` file to use that exact version.  That's what the error message is trying to communicate -- maybe we need to work on the wording.

It's required that the installed `cargo-pgrx` and the `pgrx = "x.y.z"` version Cargo.toml match.

**Comment by marcusllorente-msft** [comment]
@eeeebbbbrrrr I see. I meant to say that I did include the correct dependency version in my `.toml` file when I tried both `beta.1` and `beta.0`, but each one raised an error saying that `pgrx-0.9.8` is the currently used dependency when I already updated that. Do you have a sample `Cargo.toml` file that works using this `beta.1` version that I can use? Perhaps I am missing a tilde or a digit somewhere along the way

**Comment by eeeebbbbrrrr** [comment]
Maybe you aren't picking up the proper cargo-pgrx.  What does `cargo pgrx --version` report?

Also, all our examples work:  https://github.com/pgcentralfoundation/pgrx/tree/master/pgrx-examples

**Comment by marcusllorente-msft** [comment]
@eeeebbbbrrrr it returns: `cargo-pgrx 0.10.0-beta.1`. My `dependencies` look like this: 
`[dependencies]
pgrx = "~0.10.0-beta.1"`

**Comment by eeeebbbbrrrr** [comment]
hmm.  Change that `~` to an `=` and see what happens.

**Comment by marcusllorente-msft** [comment]
@eeeebbbbrrrr Just tried, getting the same error: `0: `pgrx-0.9.8` shouldn't be used with `cargo-pgrx-0.10.0-beta.1`, please use `pgrx = "~0.10.0-beta.1"` in your `Cargo.toml`.`

**Comment by eeeebbbbrrrr** [comment]
Can I see your entire `Cargo.toml` file?

I wonder if maybe the `pgrx-tests` dependency under the `[dev-dependencies]` block is still pointing at 0.9.8?

**Comment by eeeebbbbrrrr** [comment]
(we definitely need to clean up this error message)

**Comment by marcusllorente-msft** [comment]
@eeeebbbbrrrr Yes, just solved it. It was the `pgrx-tests` that was blocking it, had to change it to the `beta.1` version. And I tried with both a `~` and an `=`, both work fine. Thanks! But unfortunately even with the new version, it still gives me the same error with `type "x" already exists` when trying to drop and create my extension. My current workaround just to manually test is to comment out the duplicate type, but as of right now I can't do the programmed unit tests :(

**Comment by eeeebbbbrrrr** [comment]
okay.  cool.  We have enough information around that error message to be more specific, so I'll fix that.

And I'll look into this SQL generation ordering problem first thing next week.  I appreciate your tenacity and willingness to walk through some of this prickly stuff.

**Comment by marcusllorente-msft** [comment]
@eeeebbbbrrrr thanks Eric, I appreciate the help!

**Comment by eeeebbbbrrrr** [comment]
There's a PR up for this bug over in https://github.com/pgcentralfoundation/pgrx/pull/1241.

We'll get it reviewed and merged.  It'll be included in our next release, but I am unsure when that'll be.

**Comment by marcusllorente-msft** [comment]
Awesome, thanks Eric. Is there any way I can use these changes locally so I can test on my own code?
________________________________

**Comment by thomcc** [comment]
It's probably kind of painful.

1. `cargo install --git https://github.com/pgcentralfoundation/pgrx --branch develop --force cargo-pgrx`
2. Switch `version = "0.10.0-beta.1"` in your `Cargo.toml`s to `git = "https://github.com/pgcentralfoundation/pgrx", branch = "develop"`. 

However, this is fairly fragile, so you may want to switch `--branch develop` and `branch = "develop"` for `--rev 7ab42cede38c6f0826e5419ff5322f4e5789fcbf` and `rev = "7ab42cede38c6f0826e5419ff5322f4e5789fcbf"`.

You could also checkout the latest commit on the develop branch and use a `path` dependency.

Note that if you mess this up, there's no way for us to detect it and give a good error message. You're probably better off waiting for the next release (although I don't know when that will be).

**Comment by eeeebbbbrrrr** [comment]
This has been released in v0.10.0-beta.2.  https://github.com/pgcentralfoundation/pgrx/releases/tag/v0.10.0-beta.2

Enjoy!
